servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -249,97 +302,182 @@ public void onComplete() {
-                        target.onNext(oldAccumulator.finish());
+                        if (deliverOnNext(oldAccumulator, target, bSubscription, tSubscription)) {
+                            final Object nextState = maybeAccumulatorUpdater.accumulateAndGet(this,
+                                    // Keep "prev" state because we observed `itemsTerminated` or `accumulate` took
+                                    // ownership of the `holder`
+                                    toCounting(nextAccumulator), (prev, next) -> prev == holder ? next : prev);
+                            assert nextState != null;
+                            if (ItemsTerminated.class.equals(nextState.getClass())) {
+                                @SuppressWarnings("unchecked")
+                                final ItemsTerminated<T, B> it = (ItemsTerminated<T, B>) nextState;
+                                terminateIfPossible(it.accumulator, target, it.terminalNotification);
+                            }
+                        }
                         return;
                     }
                 }
             }
         }
 
-        <T, B> void itemsTerminated(final TerminalNotification terminalNotification) {
+        <T, B> void itemsTerminated(final TerminalNotification terminalNotification,
+                                    final Subscriber<? super B> target, final Cancellable bCancellable) {
             for (;;) {
                 final Object cMaybeAccumulator = maybeAccumulator;
+                assert cMaybeAccumulator != ADDING; // `accumulate` and `itemsTerminated` are sequential
                 if (cMaybeAccumulator == TERMINATED) {
                     return;
-                }
-                assert !(cMaybeAccumulator instanceof ItemsTerminated);
-
-                ItemsTerminated<T, B> itemsTerminated = new ItemsTerminated<>(cMaybeAccumulator, terminalNotification);
-                if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
-                    return;
+                } else if (cMaybeAccumulator != null &&
+                        NextAccumulatorHolder.class.equals(cMaybeAccumulator.getClass())) {
+                    // Terminated while `nextAccumulator` delivers onNext, notify that thread using ItemsTerminated
+                    // wrapper. There is no more concurrency because neither `itemsTerminated` nor `accumulate` will be
+                    // invoked again.
+                    @SuppressWarnings("unchecked")
+                    final ItemsTerminated<T, B> itemsTerminated = new ItemsTerminated<>(
+                            ((NextAccumulatorHolder<T, B>) cMaybeAccumulator).accumulator, terminalNotification);
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
+                        bCancellable.cancel();
+                        return;
```

Comment:
if errors from the subscriber thread are allowed to propagate upstream (see other comment), we may need to propagate the terminal here.

```
1. (subscriber thread) subscriber.onNext -> CAS(accumulator, ADDING)
2. (boundary thread) boundary.onNext -> CAS(ADDING, NextAccumulator)
3. (subscriber thread) target.onNext -> throws e
4. (subscriber thread) subscriber.onError(e) -> /* we should deliver the terminal event here */
```

Reply:
Not necessary in this case. If the `onNext` throws, I change the state to `Accumulator`. So, this `if` won't be used in the described use-case.

Reply:
got it, seems like we handle it when the stack unwinds iiiuc.